File: infra/lambdas/shared/onboarding-safety.ts
export type SafetyInput = {
	has_medical_condition: boolean;
	is_under_treatment: boolean;
	on_medication: boolean;
	is_pregnant_or_breastfeeding: boolean;
	has_doctor_diet_restriction: boolean;
	has_eating_disorder_history: boolean;
};

export type SafetyResult =
	| { level: "safe"; reasons: []; warnings: [] }
	| { level: "caution"; reasons: []; warnings: string[] }
	| { level: "blocked"; reasons: string[]; warnings: [] };

export function evaluateSafetyGuard(input: SafetyInput): SafetyResult {
	const blockedReasons = [
		input.is_pregnant_or_breastfeeding
			? "pregnancy_or_breastfeeding"
			: undefined,
		input.has_eating_disorder_history ? "eating_disorder_history" : undefined,
		input.has_doctor_diet_restriction ? "doctor_diet_restriction" : undefined,
	].filter((reason): reason is string => reason !== undefined);

	if (blockedReasons.length > 0) {
		return { level: "blocked", reasons: blockedReasons, warnings: [] };
	}

	const warnings = [
		input.has_medical_condition ? "medical_condition" : undefined,
		input.on_medication ? "on_medication" : undefined,
	].filter((warning): warning is string => warning !== undefined);

	if (warnings.length > 0) {
		return { level: "caution", reasons: [], warnings };
	}

	return { level: "safe", reasons: [], warnings: [] };
}

File: packages/fitness-engine/src/fitness_engine/onboarding_safety.py
"""Onboarding Flow (Plan 07) 用の bool subset Safety adapter。

既存 `fitness_engine.safety` は Plan 生成用の broader contract を扱うため、
Onboarding 画面の二重防御とは分離して mirror 実装する。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal


@dataclass(frozen=True)
class OnboardingSafetyInput:
    has_medical_condition: bool
    is_under_treatment: bool
    on_medication: bool
    is_pregnant_or_breastfeeding: bool
    has_doctor_diet_restriction: bool
    has_eating_disorder_history: bool


@dataclass(frozen=True)
class OnboardingSafetyResult:
    level: Literal["safe", "caution", "blocked"]
    reasons: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


def evaluate_onboarding_safety_guard(
    input_: OnboardingSafetyInput,
) -> OnboardingSafetyResult:
    blocked_reasons: list[str] = []
    if input_.is_pregnant_or_breastfeeding:
        blocked_reasons.append("pregnancy_or_breastfeeding")
    if input_.has_eating_disorder_history:
        blocked_reasons.append("eating_disorder_history")
    if input_.has_doctor_diet_restriction:
        blocked_reasons.append("doctor_diet_restriction")

    if blocked_reasons:
        return OnboardingSafetyResult(level="blocked", reasons=blocked_reasons)

    warnings: list[str] = []
    if input_.has_medical_condition:
        warnings.append("medical_condition")
    if input_.on_medication:
        warnings.append("on_medication")

    if warnings:
        return OnboardingSafetyResult(level="caution", warnings=warnings)

    return OnboardingSafetyResult(level="safe")
